File: android/src/main/java/com/jjkit/drawableview/DrawableViewManager.java
package com.jjkit.drawableview;

import android.graphics.Color;
import android.graphics.Paint;
import android.util.TypedValue;

import androidx.annotation.NonNull;

import com.facebook.react.bridge.Dynamic;
import com.facebook.react.bridge.ReadableArray;
import com.facebook.react.bridge.ReadableMap;
import com.facebook.react.uimanager.ThemedReactContext;
import com.facebook.react.uimanager.ViewGroupManager;
import com.facebook.react.uimanager.annotations.ReactProp;


class DrawableViewManager extends ViewGroupManager<DrawableView> {


    @Override
    public String getName() {
        return "JJSDrawable";
    }

    @Override
    protected DrawableView createViewInstance(@NonNull ThemedReactContext reactContext) {
        return new DrawableView(reactContext);
    }

    @ReactProp(name = "d")
    public void setD(DrawableView view, String v){
        String p = v == null ? "" :  v;
        view.setD(p);
    }
    @ReactProp(name = "viewBox")
    public void setViewBox(DrawableView view, ReadableArray viewBox){
        float [] v = {0f,0f,-1f,-1f};
        if(viewBox != null) {
            v[0] = (float) viewBox.getDouble(0);
            v[1] = (float) viewBox.getDouble(1);
            v[2] = (float) viewBox.getDouble(2);
            v[3] = (float) viewBox.getDouble(3);
        }
        view.setViewBox(v);
    }
    @ReactProp(name = "align")
    public void setAlign(DrawableView view, String v){
        String p = v == null ? "xMidYMid" :  v;
        view.setAlign(p);
    }
    @ReactProp(name = "aspect")
    public void setAspect(DrawableView view, String v){
        int p = v == null ? SVGViewBox.MOS_MEET :  (v.equals("slice") ? SVGViewBox.MOS_SLICE :  (v.equals("none") ? SVGViewBox.MOS_NONE : SVGViewBox.MOS_MEET ) );
        view.setAspect(p);
    }

    @ReactProp(name = "translateZ",defaultFloat = 0f)
    public void setTranslateZ(DrawableView view ,float v) {
        view.setTranslateZ(v);
    }
    @ReactProp(name = "opacity",defaultFloat = 1f)
    public void setOpacity(DrawableView view ,float v) {
        view.setOpacity(v);
    }

    @ReactProp(name = "fill",defaultInt = Color.BLACK)
    public void setFill(DrawableView view , int v) {
        view.setFill(v);
    }
    @ReactProp(name = "fillRule")
    public void setFillRule(DrawableView view ,String v) {
        view.setFillRule(v == null ? "none" : v);

    }
    @ReactProp(name = "fillOpacity",defaultFloat = 1f)
    public void setFillOpacity(DrawableView view ,float v) {
        view.setFillOpacity(v);

    }
    @ReactProp(name = "stroke",defaultInt = Color.TRANSPARENT)
    public void setStroke(DrawableView view , int v) {
        view.setStroke(v);

    }
    @ReactProp(name = "strokeOpacity",defaultFloat = 1f)
    public void setStrokeOpacity(DrawableView view ,float v) {
        view.setStrokeOpacity(v);

    }

    @ReactProp(name = "strokeWidth",defaultFloat = 1f)
    public void setStrokeWith(DrawableView view ,float v) {
        view.setStrokeWith(v);

    }
    @ReactProp(name = "strokeCap")
    public void setStrokeCap(DrawableView view ,String v) {
        view.setStrokeCap(v == null ? "none" : v);

    }

    @ReactProp(name = "strokeJoin")
    public void setStrokeJoin(DrawableView view ,String v) {
        view.setStrokeJoin(v == null ? "none" : v);

    }
    @ReactProp(name = "strokeMiter",defaultFloat = 4f)
    public void setStrokeMiter(DrawableView view ,float v) {
        view.setStrokeMiter(v);

    }

    @ReactProp(name = "strokeStart",defaultFloat = 0f)
    public void setStrokeStart(DrawableView view ,float v) {
        view.setStrokeStart(v);

    }
    @ReactProp(name = "strokeEnd",defaultFloat = 1f)
    public void setStrokeEnd(DrawableView view ,float v) {
        view.setStrokeEnd(v);

    }

    @ReactProp(name = "shadow",defaultInt = Color.BLACK)
    public void setShadow(DrawableView view ,int v) {
        view.setShadow(v);

    }
    @ReactProp(name = "shadowOpacity",defaultFloat = 0f)
    public void setShadowOpacity(DrawableView view ,float v) {
        view.setShadowOpacity(v);

    }
    @ReactProp(name = "shadowRadius",defaultFloat = 2f)
    public void setShadowRadius(DrawableView view ,float v) {
        view.setShadowRadius(v);

    }

    @ReactProp(name = "shadowOffset",defaultFloat = 2f)
    public void setShadowOffset(DrawableView view , float v) {
        view.setShadowOffset(v);
    }
    @ReactProp(name = "shadowOffsetX",defaultFloat = 2f)
    public void setShadowOffsetX(DrawableView view , float v) {
        view.setShadowOffsetX(v);
    }
    @ReactProp(name = "shadowOffsetY",defaultFloat = 2f)
    public void setShadowOffsetY(DrawableView view , float v) {
        view.setShadowOffsetY(v);
    }
    @ReactProp(name = "shadowPercentageValue",defaultBoolean  = false)
    public void setShadowPercentageValue(DrawableView view , boolean v) {
        view.setShadowPercentageValue(v);
    }

    //MARK: Transform

    @ReactProp(name = "transX",defaultFloat = 0f)
    public void setTransX(DrawableView view, float v) {
        view.setTransX(v);
    }
    @ReactProp(name = "transY",defaultFloat = 0f)
    public void setTransY(DrawableView view, float v) {
        view.setTransY(v);
    }
    @ReactProp(name = "transPercentageValue",defaultBoolean = false)
    public void setTransPercentageValue(DrawableView view, boolean v) {
        view.setTransPercentageValue(v);
    }

    @ReactProp(name = "rot",defaultFloat = 0f)
    public void setRot(DrawableView view , float v) {
        view.setRot(v);
    }
    @ReactProp(name = "rotO",defaultFloat = 0f)
    public void setRotO(DrawableView view , float v) {
        view.setRotO(v);
    }
    @ReactProp(name = "rotOx",defaultFloat = 0f)
    public void setRotOx(DrawableView view , float v) {
        view.setRotOx(v);
    }
    @ReactProp(name = "rotOy",defaultFloat = 0f)
    public void setRotOy(DrawableView view , float v) {
        view.setRotOy(v);
    }
    @ReactProp(name = "rotPercentageValue",defaultBoolean  = false)
    public void setRotPercentageValue(DrawableView view , boolean v) {
        view.setRotPercentageValue(v);
    }

    @ReactProp(name = "sc",defaultFloat = 1f)
    public void setSc(DrawableView view, float v) {
        view.setSc(v);
    }
    @ReactProp(name = "scX",defaultFloat = 1f)
    public void setScX(DrawableView view, float v) {
        view.setScX(v);
    }
    @ReactProp(name = "scY",defaultFloat = 1f)
    public void setScY(DrawableView view, float v) {
        view.setScY(v);
    }
    @ReactProp(name = "scO",defaultFloat = 0f)
    public void setScO(DrawableView view, float v) {
        view.setScO(v);
    }
    @ReactProp(name = "scOx",defaultFloat = 0f)
    public void setScOx(DrawableView view, float v) {
        view.setScOx(v);
    }
    @ReactProp(name = "scOy",defaultFloat = 0f)
    public void setScOy(DrawableView view, float v) {
        view.setScOy(v);
    }
    @ReactProp(name = "scPercentageValue",defaultBoolean = false)
    public void setScPercentageValue(DrawableView view, boolean v) {
        view.setScPercentageValue(v);
    }
 
   
}

File: android/src/main/java/com/jjkit/drawableview/DrawableViewManager.kt
package com.jjkit.drawableview

import android.content.Context
import android.graphics.Color
import android.graphics.Paint
import android.util.TypedValue
import android.view.View
import com.facebook.react.bridge.ReadableMap
import com.facebook.react.uimanager.annotations.ReactProp
import com.facebook.react.uimanager.*
import com.jjlf.jjkit_utils.drawables.JJDrawable


class DrawableViewManager : ViewGroupManager<DrawableView>() {

    override fun getName(): String {
        return "Drawable"
    }

    override fun createViewInstance(reactContext: ThemedReactContext): DrawableView {
        val v = DrawableView(reactContext as Context)
        v.setLayerType(View.LAYER_TYPE_SOFTWARE,null)
        return v
    }

    @ReactProp(name = "path")
    fun setPath(view: DrawableView, p:ReadableMap?){


        val d = try { p!!.getString("d")!! }catch(e: Exception) { "" }
        val viewBox = try { p!!.getArray("viewBox") }catch(e: Exception) { null }
        val aspect = try { p!!.getString("aspect")!! }catch(e: Exception) { "none" }
        val align = try { p!!.getString("align")!! }catch(e: Exception) { "none" }

        val a  = when(aspect) {
            "meet" ->  JJDrawable.VIEW_BOX_MEET
            "slice" -> JJDrawable.VIEW_BOX_SLICE
            else -> JJDrawable.VIEW_BOX_NONE
        }

        val v = floatArrayOf(0f,0f,0f,0f)
        viewBox?.let {
            v[0] = it.getDouble(0).toFloat()
            v[1] = it.getDouble(1).toFloat()
            v[2] = it.getDouble(2).toFloat()
            v[3] = it.getDouble(3).toFloat()
        }
        val mDrawable = view.getDrawable()
        mDrawable.setSvgPath(d,view.resources.displayMetrics.density, v,align,a)
        mDrawable.invalidateSelf()


    }
    @ReactProp(name = "pathRotation",defaultFloat = 0f)
    fun setPathRotation(view: DrawableView, r:Float){
        val mDrawable = view.getDrawable()
        mDrawable.setPathRotation(r)
        mDrawable.invalidateSelf()
    }
    @ReactProp(name = "pathScale")
    fun setPathScale(view: DrawableView, v:ReadableMap?){
        val mDrawable = view.getDrawable()
        val x = try { v!!.getDouble("x") }catch(e: Exception) { 1.0 }.toFloat()
        val y = try { v!!.getDouble("y") }catch(e: Exception) { 1.0 }.toFloat()
        mDrawable.setPathScale(x,y)
        mDrawable.invalidateSelf()
    }
    @ReactProp(name = "pathTranslation")
    fun setPathTranslation(view: DrawableView, v:ReadableMap?){
        val dx = try { v!!.getDouble("dx") }catch(e: Exception) { 0.0 }.toFloat()
        val dy = try { v!!.getDouble("dy") }catch(e: Exception) { 0.0 }.toFloat()
        val per = try { v!!.getBoolean("percentageValue") }catch(e: Exception) { false }
        val mDrawable = view.getDrawable()
        if(per){
            mDrawable.setPathTranslation(dx,dy,0f,0f)
        }else{
            mDrawable.setPathTranslation(toDip(dx,view),toDip(dy,view))
        }
        mDrawable.invalidateSelf()
    }

    @ReactProp(name = "shadowColor",defaultInt = Color.BLACK)
    override fun setShadowColor(view: DrawableView, v:Int){
        val mDrawable = view.getDrawable()
        mDrawable.setShadowColor(v)
        mDrawable.invalidateSelf()
    }
    @ReactProp(name = "shadowOffset")
    fun setShadowOffset(view: DrawableView, v:ReadableMap?){
        val x = try { v!!.getDouble("x") }catch(e: Exception) { 0.0 }.toFloat()
        val y = try { v!!.getDouble("y") }catch(e: Exception) { 0.0 }.toFloat()
        val mDrawable = view.getDrawable()
        mDrawable.setShadowOffset(toDip(x,view),toDip(y,view))
        mDrawable.invalidateSelf()
    }
    @ReactProp(name = "shadowOpacity",defaultFloat = 0f)
    fun setShadowOpacity(view: DrawableView,v:Float){
        val mDrawable = view.getDrawable()
        mDrawable.setShadowOpacity(v)
        mDrawable.invalidateSelf()
    }
    @ReactProp(name = "shadowRadius",defaultFloat = 1f)
    fun setShadowRadius(view: DrawableView,v:Float){
        val mDrawable = view.getDrawable()
        mDrawable.setShadowRadius(toDip(v,view))
        mDrawable.invalidateSelf()
    }
    @ReactProp(name = "strokeWidth",defaultFloat = 0f)
    fun setStrokeWidth(view: DrawableView,v:Float){
        val mDrawable = view.getDrawable()
        mDrawable.setStrokeWidth(toDip(v,view))
        mDrawable.invalidateSelf()
    }
    @ReactProp(name = "strokeColor",defaultInt = Color.BLACK)
    fun setStrokeColor(view: DrawableView,v:Int){
        val mDrawable = view.getDrawable()
        mDrawable.setStrokeColor(v)
        mDrawable.invalidateSelf()
    }
    @ReactProp(name = "strokeStart",defaultFloat = 0f)
    fun setStrokeStart(view: DrawableView,v:Float){
        val mDrawable = view.getDrawable()
        mDrawable.setStrokeStart(v)
        mDrawable.invalidateSelf()
    }
    @ReactProp(name = "strokeEnd",defaultFloat = 1f)
    fun setStrokeEnd(view: DrawableView,v:Float){
        val mDrawable = view.getDrawable()
        mDrawable.setStrokeEnd(v)
        mDrawable.invalidateSelf()
    }
     @ReactProp(name = "strokeCap")
    fun setStrokeCap(view: DrawableView,v:String?){
        val mDrawable = view.getDrawable()
        val cap = when(v ?: "butt"){
                "round" -> Paint.Cap.ROUND
                "square" -> Paint.Cap.SQUARE
                 else -> Paint.Cap.BUTT
        }
        mDrawable.setStrokeCap(cap)
        mDrawable.invalidateSelf()
    }
    @ReactProp(name = "strokeJoin")
    fun setStrokeJoin(view: DrawableView,v:String?){
        val mDrawable = view.getDrawable()
        val join = when(v ?: "miter"){
            "round" -> Paint.Join.ROUND
            "bevel" -> Paint.Join.BEVEL
            else -> Paint.Join.MITER
        }
        mDrawable.setStrokeJoin(join)
        mDrawable.invalidateSelf()
    }
    @ReactProp(name = "strokeMiter",defaultFloat = 4f)
    fun setStrokeMiter(view: DrawableView,v:Float){
        val mDrawable = view.getDrawable()
        mDrawable.setStrokeMiter(v)
        mDrawable.invalidateSelf()
    }
    @ReactProp(name = "pathInset")
    fun setPathInset(view: DrawableView,v:ReadableMap?){
        val mDrawable = view.getDrawable()
        val x = try { v!!.getDouble("x") }catch(e: Exception) { 0.0 }.toFloat()
        val y = try { v!!.getDouble("y") }catch(e: Exception) { 0.0 }.toFloat()
        mDrawable.setInset(x,y)
        mDrawable.invalidateSelf()
    }
    @ReactProp(name = "fillColor", defaultInt = Color.TRANSPARENT)
    fun setFillColor(view: DrawableView,v:Int){
        val mDrawable = view.getDrawable()
        mDrawable.setFillColor(v)
        mDrawable.invalidateSelf()
    }
    @ReactProp(name = "bgColor",defaultInt = Color.TRANSPARENT)
    fun setBgColor(view: DrawableView,v:Int){
        val mDrawable = view.getDrawable()
        mDrawable.setBackgroundColor(v)
        mDrawable.invalidateSelf()
    }


    private fun toDip(num:Float, view: DrawableView): Float{
        return TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_DIP,num,view.resources.displayMetrics)
    }
}